types/src/keyless/proof_simulation.rs

```diff
@@ -0,0 +1,259 @@
+            circuit,
+            alpha,
+            beta,
+            gamma,
+            delta,
+            g1_generator,
+            g2_generator,
+            rng,
+        ).unwrap();
+
+        Ok((
+        ProvingKeyWithTrapdoor {
+            gamma_abc_g1: pk.vk.gamma_abc_g1.clone(),
+            alpha,
+            beta,
+            delta,
+            gamma,
+            g1: g1_generator.into_affine(),
+            g2: g2_generator.into_affine(),
+        },
+        pk.vk.clone()))
+    }
+
+    /// Create a Groth16 proof that is zero-knowledge using the provided
+    /// R1CS-to-QAP reduction.
+    /// This method samples randomness for zero knowledges via `rng`.
+    #[inline]
+    pub fn create_random_proof_with_trapdoor(
+        public_inputs: &[E::ScalarField],
+        pk: &ProvingKeyWithTrapdoor<E>,
+        rng: &mut impl Rng,
+    ) -> R1CSResult<Proof<E>>
+    where
+    {
+        let a = E::ScalarField::rand(rng);
+        let b = E::ScalarField::rand(rng);
+
+        Self::create_proof_with_trapdoor(pk, a, b, public_inputs)
+    }
+
+    /// Creates proof using the trapdoor
+    #[cfg(test)]
+    pub fn create_proof_with_trapdoor(
+        pk: &ProvingKeyWithTrapdoor<E>,
+        a: E::ScalarField,
+        b: E::ScalarField,
+        input_assignment: &[E::ScalarField],
+    ) -> R1CSResult<Proof<E>> {
+        let public_inputs = input_assignment;
+        let mut g_ic = pk.gamma_abc_g1[0].into_group();
+        for (i, b) in public_inputs.iter().zip(pk.gamma_abc_g1.iter().skip(1)) {
+            g_ic.add_assign(&b.mul_bigint(i.into_bigint()));
+        }
+        g_ic = g_ic * pk.gamma;
+
+        let delta_inverse = pk.delta.inverse().unwrap();
+        let ab = a * b;
+        let alpha_beta = pk.alpha * pk.beta;
+
+        let g1_ab = pk.g1 * ab;
+        let g1_alpha_beta = pk.g1 * alpha_beta;
+        
+        let g1_a = pk.g1 * a;
+        let g2_b = pk.g2 * b;
+
+        let g1_c = (g1_ab - g1_alpha_beta - g_ic) * delta_inverse; 
+
+        Ok(Proof {
+            a: g1_a.into_affine(),
+            b: g2_b.into_affine(),
+            c: g1_c.into_affine(),
+        })
+    }
+}
+
+
+
+/// Generates a trapdoor proving and verifiying key pair intended for proof simulation, in addition to a vector of public inputs, from
+/// circom-generated .r1cs and .wasm files, and a .json file containing the public inputs. To be
+/// used to update `test_prove_and_verify` after circuit changes occur
+/// WARNING: The files referenced in this function are not present in this repo and msut be added
+/// manually
+// TODO: There is a version conflict between this piece of code and `aptos-dkg` with the
+// `num-bigint` dependency. Ideally we resolve this by updating `aptos-dkg` to `num-bigint` version
+// `0.4.5`
+/*#[allow(dead_code)]
+fn generate_keys_and_inputs<E: Pairing>() {
+    let cfg = CircomConfig::<E>::new(
+        "./circuit-files/keyless_main.wasm",
```

Comment:
Change this to take the paths as args. Define the paths as globals via in `include!`

Reply:
I don't really see the use of `include!` here, what do you have in mind? I've changed the funciton to take in the paths as parameters

Reply:
That's good that the function takes the path as a parameter, but how will devs know where the keyless circuit path is in our codebase?

I am saying you should define a static/global for the circuit's path by using `include!`.

I would also propose you add a `keyless_setup_with_trapdoor(rng)` function which just calls `circuit_specific_setup_with_trapdoor` with the paths from the static variable.

Reply:
Should we include the keyless circuit in our codebase? I did not include it due to its large size

Reply:
What do you mean? It's already included: https://github.com/aptos-labs/aptos-core/tree/main/keyless/circuit

Reply:
Or do you mean the compiled R1CS?

Reply:
Yes I mean the compiled R1CS, which is what the function requires a path to (in addition to the compiled `.wasm` file)

Reply:
How big is it?

If it's too big, this may be a good reason to solely rely on `circuit_agnostic_specific_setup_with_trapdoor`

Reply:
About 345MB.

I don't see a reason this suggests we should only use the circuit agnostic function when we can just hardcode the VK - in the worst case we hardcode the wrong VK and we're reduced to what is essentially a circuit agnostic version of the code anyways.

Reply:
Yeah, it's just that... you will generate this hardcoded VK and likely won't leave instructions anywhere on how to regenerate it 😆 

So when it comes to regenerating next time, we'll waste a lot of time.

So unless you have a script with a README to regenerate this VK and its trapdoor, I propose we use the circuit agnostic thing, which will always work.

We gotta think of maintainability of this code also.

Reply:
Right. A README to install the right dependencies for that script to run might also be needed.

Reply:
Probably easier to just create a virtual environment in the script. A README would be good to know to even run it though